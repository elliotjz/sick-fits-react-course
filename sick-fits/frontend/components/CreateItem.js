import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import React, { Component } from 'react'
import Router from 'next/router'

import Error from './ErrorMessage'
import Form from './styles/Form'
import formatMoney from '../lib/formatMoney'

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: 'Cool shoes',
    description: 'Some rando desc',
    image: '',
    largeImage: '',
    price: 1000,
  }

  handleChange = e => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({ [name]: val })
  }

  uploadFile = async e => {
    const files = e.target.files
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', 'sickfits')

    const res = await fetch(
      'https://api.cloudinary.com/v1_1/dqtw0enxk/image/upload',
      {
        method: 'POST',
        body: data,
      }
    )
    const file = await res.json()

    if (file.error) {
      this.setState({
        image: '',
        largeImage: '',
      })
    } else {
      this.setState({
        image: file.secure_url,
        largeImage: file.eager[0].secure_url,
      })
    }
  }

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              // Stop the form from submitting
              e.preventDefault()
              // Cal the mutation
              const res = await createItem()
              // Change them to the singe item page
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              })
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.uploadFile}
                />
              </label>

              {this.state.image && (
                <img
                  width="200px"
                  src={this.state.image}
                  alt="Upload Preview"
                />
              )}

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter a Description"
                  required
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>

              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
