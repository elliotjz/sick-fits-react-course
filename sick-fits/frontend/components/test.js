import React, { Component } from 'react'

class MyComponent extends Component {
  render() {
    return <Something>Hello</Something>
  }
}

export default MyComponent
