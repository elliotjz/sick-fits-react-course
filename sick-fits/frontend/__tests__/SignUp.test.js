import { mount } from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import { MockedProvider } from 'react-apollo/test-utils'
import { ApolloConsumer } from 'react-apollo'

import SignUp, { SIGNUP_MUTATION } from '../components/SignUp'
import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser } from '../lib/testUtils'

function type(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: { name, value },
  })
}
const me = fakeUser()

const mocks = [
  // SignUp mock mutation
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        email: me.email,
        name: me.name,
        password: '123',
      },
    },
    result: {
      data: {
        signup: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  // Current user query mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me } },
  },
]

describe('<SignUp/>', () => {
  it('renders and matches the snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    )
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot()
  })

  it('calls the mutation properly', async () => {
    let apolloClient
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client
            return <SignUp />
          }}
        </ApolloConsumer>
      </MockedProvider>
    )

    await wait()
    wrapper.update()
    type(wrapper, 'name', me.name)
    type(wrapper, 'email', me.email)
    type(wrapper, 'password', '123')
    wrapper.update()
    wrapper.find('form').simulate('submit')
    await wait()

    // Query the user out of the apollow client
    const user = await apolloClient.query({ query: CURRENT_USER_QUERY })

    expect(user.data.me).toMatchObject(me)
  })
})
