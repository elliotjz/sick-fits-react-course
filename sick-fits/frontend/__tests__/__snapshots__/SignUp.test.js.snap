// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> renders and matches the snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 fCpNnG"
  method="POST"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Up For an Account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
